### PR TITLE
[ENH] Revision Forecasting Benchmarking Module

### PR DESCRIPTION
**Pull request by benHeid**
#### Reference Issues/PRs

# Benchmarking Requirements:
These are the requirements, I tried to extract from the #5196 and #4873. Feel free to add if forgot some here. In the following, I try to mention all of them and answer how I tried to solve them
## Registry related: 
* Model registry should be able to add multiple estimators add once:
* Registry should be storable and loadable.
* Registry Should manage the fitted models 
  * This need changes to evaluate. Evaluate needs to return the fitted models.

* Registry wide modifiers (e.g. reconcilers) (I do not understand it).

## Result Storages
* Decouple storage approach (strategy pattern)
  * Based on the file ending, of the path provided in run, a different backend is selected

* the file run provided to run should be not only csv. 
   * JSON and (Perhaps Parquet) are currently possible
* Purely in memory if no file is provided:
  * Not implemented for now. Is this useful?

## Aggregations
* Result should return Intermediate Results (GT, Prediction, Training Data). To enable user better analysis
  * This is implemented. Each fold has the training, testing data as well as the prediction

* aggregation should be flexible over folds, hierarchical levels, and test data indices.
  * This is supported by the metrics. Passing such metrics will lead to correspondingly aggregated means over all folds.

## Syntax
* benchmark class could be polymorphic, i.e., dispatch to classifier, forecaster benchmark etc.
  * NOT IMPLEMENT FOR NOW
* estimator_id should be unrestricted
  * All strings are allowed -> Uses class_name if not provided.
   * OPEN how to handle conflicts
* add_estimator interface is unnecessarily complex (shorten, automatic clone, allow fitted modules)
  * I am in favor of the long name. Clone happens in evaluate. The usage of fitted models would probably require changes to evaluate
## Task Creation
* Beside data loade, also in memory and new Dataset object should be allowed.
   * Dataset Object, and sktime compatible in memory data formats are still open.

# Architectural overview.
* The main class is the benchmark object. 
   * It stores all tasks and estimators that should be benchmarked in the registry implementation.
    * public methods are the same as for old benchmarking module
    * run executes all combination of task and estimator. Result are stored in ResultObjects1
  * Results are managed via multiple data classes.
    * Better decoupling of benchmarking and the management of the data.
    * Aggregation tasks etc. can be done there. Makes the benchmarking module itself easier maintainable.

Pining @fkiraly and @felipeangelimvieira explicitly, since I think you are working on some related topics. Please challenge my current design. 

To See how the user could use the framework, check the notebook.

### CODE REVIEW

File: sktime/benchmarking/benchmarking_dataclasses.py [hunk 1]
```diff
@@ -0,0 +1,430 @@
...
+            self.stds.append(ScoreResult(name, np.std(score, axis=0)))
+
+
+@dataclass
+class BenchmarkingResults:
+    """Results of a benchmarking run.
+
+    Parameters
+    ----------
+    results : list of ResultObject
```
Comment (fkiraly) [inline sktime/benchmarking/benchmarking_dataclasses.py:146]: results are indexed by task and model ID, so a `dict` would be far more efficient for retrieval. `contains` would then just be a query of the keys.

File: sktime/benchmarking/benchmarking_dataclasses.py [hunk 2]
```diff
@@ -0,0 +1,430 @@
...
+            results_df = pd.read_parquet(self.path)
+            results = []
+            for ix, row in results_df.iterrows():
+                folds = {}
+                for fold in row["folds"]:
+                    scores = []
+                    for score in fold["scores"]:
+                        scores.append(ScoreResult(score["name"], score["score"]))
+                    if "ground_truth" in fold:
+                        ground_truth = pd.Series(fold["ground_truth"])
```
Comment (fkiraly) [inline sktime/benchmarking/benchmarking_dataclasses.py:387]: ground truth etc are not always `pd.Series`, e.g., in the multivariate case?

**Review by fkiraly — CHANGES_REQUESTED** [review]
Some comments.

### Data class model

* I think a task is not fully defined by the parameters in it, as we could refit or update. I think the full specification requires passing the args of `evaluate` excluding the backend ones, e.g., `strategy` and `cv_X` as well.
* I would make the "data" field more general - allowing passing `X` and `y` as in `evaluate`, but also lazy variants such as the loader.
* `ModelToTest` looks like it could be a `dict` or `DataFrame` row/col
* `ScoreResult` looks like it could be a `dict` or `DataFrame` row/col
* `ResultObject` looks like it coudl be a `dict`, indexed by `model_id` and `task_id`

## conceptual model

The benchmark does not seem to allow for evaluation by horizon? Should we allow that?

Similarly, aggregation/disaggregation modes for panel data might be worth thinking about, at leats with a view towards upwards compatibility.

File: sktime/benchmarking/benchmarking_dataclasses.py [hunk 3]
```diff
@@ -0,0 +1,430 @@
...
+def _asdict_inner(obj, dict_factory, pd_orient):
+    # Copied from dataclasses._asdict_inner and slightly modified
+    if hasattr(type(obj), "__dataclass_fields__"):
+        result = []
+        for f in fields(obj):
+            value = _asdict_inner(getattr(obj, f.name), dict_factory, pd_orient)
+            result.append((f.name, value))
+        return dict_factory(result)
+    elif isinstance(obj, tuple) and hasattr(obj, "_fields"):
+        # obj is a namedtuple.  Recurse into it, but the returned
```
Comment (fkiraly) [inline sktime/benchmarking/benchmarking_dataclasses.py:201]: is the problem that you are trying to solve the same that `get_params` in `skbase` solves?

**Comment by benHeid** [comment]
Are answering the questions first. Regarding the dataclass I new to check. 

> The benchmark does not seem to allow for evaluation by horizon? Should we allow that?

Should be possible when the dispatch in call in the metrics is merged. Then the user can control this by adding a metric with the parameter by_index. I think this is more elegant than adding an additional parameter here. 


> Similarly, aggregation/disaggregation modes for panel data might be worth thinking about, at leats with a view towards upwards compatibility.

Agree. Maybe, I need multiple tasks objects internally to make this pretty.

**Comment by fkiraly** [comment]
> Should be possible when the dispatch in call in the metrics is merged. Then the user can control this by adding a metric with the parameter by_index. I think this is more elegant than adding an additional parameter here.

I see - but then the `score`-s would not be floats but could be `pandas` objects?

**Comment by benHeid** [comment]
> I see - but then the `score`-s would not be floats but could be `pandas` objects?

Yes as in the case of hierarchical data are used with the corresponding metric parameter.

**Comment by fkiraly** [comment]
ok, then custom containers do make sense, although I think the point still holds that they could be `dict`-s?

**Review by fkiraly — CHANGES_REQUESTED** [review]
Reviewing this, main comments:

* classes vs dicts, as above
* new point: is it a good idea to add `run` to the `ForecastingBenchmark`? Currently this is in the `BaseBenchmark`.

I would propose to leave it there, but add evaluator classes that are specific to the scientific type. I will try to experiment with this in a new PR.

**Comment by benHeid** [comment]
> * classes vs dicts, as above

would you replace all of the data classes by dicts? It might be that `dicts` are more efficient, but I suppose that this overhead here is negligible and on the other hand dataclasses are easier  to understand and also better testable. However, I see that perhaps the number of dataclasses can be reduced. 
 
> * new point: is it a good idea to add `run` to the `ForecastingBenchmark`? Currently this is in the `BaseBenchmark`.
> 
> I would propose to leave it there, but add evaluator classes that are specific to the scientific type. I will try to experiment with this in a new PR.

Mhm, I think having specific evaluator would be nice. But regarding reducing the dependencies of the different pull request, should we continue with the current approach and change it later?

**Comment by fkiraly** [comment]
> Mhm, I think having specific evaluator would be nice. But regarding reducing the dependencies of the different pull request, should we continue with the current approach and change it later?

I'd rather not have a fourth entirely separate benchmarking framework in the code base - I think we need to gradually change the current `ForecastingBenchmark` and ensure it remains downwards compatible, for a while at least.

I like your approach, I would just recommend to start from the current file and the current tests. Optionally, we could cut into smaller stepwise changes.

**Comment by fkiraly** [comment]
> > * classes vs dicts, as above
> 
> would you replace all of the data classes by dicts?

Not universally, let me go through one-by-one:

* `TaskObject` - I would change this to something like the evaluator class in #7913. This ensures that the API is not scitype specific.
* `ModelToTest` and `ScoreResult` feel like `dict`, since they have two fields, and one is an index, and they are always used in a `list`. Access in this structure is also much less efficient than in the `dict`.
* `SktimeRegistry` feels like a `dict`, it does not add much on top of `self.entities`. The `register` method maps to the write operation
* `BenchmarkingResults` and `ResultsObject` seem fine, although it should be possible to save sequentially as jobs complete, rather then only when all jobs have completed. I also wonder whether they could be "flattened" or "merged", but I have no major reservations here.
* `BenchmarkingResults` *could* also be a `dict`, indexed by a tuple `(model_id, task_id)`.

In terms of PR strategy, I think there are *two big changes* that could be separated into two PR:

* restructuring the current benchmarking logic into classes, containers, etc
* adding backend data handlers

The latter imo is much clearer design-wise, of course it depends on `ResultsObject` at the moment.

The one remark I have about the data handling, I would make `get_storage_backend` extensible, i.e., it should not need changing if a new data handler class gets added. I would adopt the same pattern as for `deep_equals` or the `datatypes` module, where the "check whether the plugin applies" is localized in the handler class, and the `get_storage_backend` method collects all the plugins.

**Comment by benHeid** [comment]
> * `TaskObject` - I would change this to something like the evaluator class in [[ENH] benchmarking framework improvement - refactor and validator design #7913](https://github.com/sktime/sktime/pull/7913). This ensures that the API is not scitype specific.
Make sense.

> * `ModelToTest` and `ScoreResult` feel like `dict`, since they have two fields, and one is an index, and they are always used in a `list`. Access in this structure is also much less efficient than in the `dict`.
Can be rewritten as a dict. Which would limit extensibility in case there are more fields we would like to add.

> * `SktimeRegistry` feels like a `dict`, it does not add much on top of `self.entities`. The `register` method maps to the write operation

I would recommend to keep it. Since this would enable us to handle cases where the user adds an estimator two times with the same id more easily. In general, I would keep the classes whenever we are defining an operation on it since this makes the code structure cleaner, since the functionality is always part of the object. 

> * `BenchmarkingResults` and `ResultsObject` seem fine, although it should be possible to save sequentially as jobs complete, rather then only when all jobs have completed. I also wonder whether they could be "flattened" or "merged", but I have no major reservations here.

BenchmarkingResult is implementing already a `save` method. Thus, it should be always storable. I.e., after each new evaluation a result could be stored. 

> * `BenchmarkingResults` _could_ also be a `dict`, indexed by a tuple `(model_id, task_id)`.

By similar reasons as for sktimeRegistry, I would not do it. In case of the results, we have operations that will be performed always on results, as storing or transforming them. Thus, I would not change them to a dict.

> In terms of PR strategy, I think there are _two big changes_ that could be separated into two PR:
> 
> * restructuring the current benchmarking logic into classes, containers, etc
> * adding backend data handlers
> 
> The latter imo is much clearer design-wise, of course it depends on `ResultsObject` at the moment.

Sou you would keep the current test set and just replace the existing Benchmark by that one and assume that we are backward compatible if all of the tests work without being changed? 


> The one remark I have about the data handling, I would make `get_storage_backend` extensible, i.e., it should not need changing if a new data handler class gets added. I would adopt the same pattern as for `deep_equals` or the `datatypes` module, where the "check whether the plugin applies" is localized in the handler class, and the `get_storage_backend` method collects all the plugins.

I agree, could we solve this by introducing the base class design here too and make the storage Backends discoverable?

**Comment by fkiraly** [comment]
>> `ModelToTest` and `ScoreResult` feel like dict, since they have two fields, and one is an index, and they are always used in a list. Access in this structure is also much less efficient than in the dict.

> Can be rewritten as a dict. Which would limit extensibility in case there are more fields we would like to add.

I would do that once there are more fields to add, and leave the `dict` but just add a `dataclass`-like container for multiple fields.

> By similar reasons as for SktimeRegistry, I would not do it. In case of the results, we have operations that will be performed always on results, as storing or transforming them.

Ok, I do not have a strong opinion about these. Though, can we make the classes private to ensure refactoring is easier? They are used only internally anyway.

> I would recommend to keep it. Since this would enable us to handle cases where the user adds an estimator two times with the same id more easily

Sure - I would handle it though consistenly with `_HeterogenousMetaEstimator`, where a name clash is resolved by appending `_1` and `_2` etc.

> Sou you would keep the current test set and just replace the existing Benchmark by that one and assume that we are backward compatible if all of the tests work without being changed?

Yes, it if is possible without major work - the public interface should optimally stay the same, or we have to go through deprecation. If we change the interface, we should carefully think about why and how to deprecate.

> I agree, could we solve this by introducing the base class design here too and make the storage Backends discoverable?

I would make them private and only extensible for now, not publicly discoverable, that gives us more flexibility to change things while the API is still maturing.

Side note, I would also collect them all in a single file or folder with individual backends in a private submodule.

**Comment by fkiraly** [comment]
(oops, mouse slipped on "close", reopened)

**Comment by benHeid** [comment]
> > > `ModelToTest` and `ScoreResult` feel like dict, since they have two fields, and one is an index, and they are always used in a list. Access in this structure is also much less efficient than in the dict.
> 
> > Can be rewritten as a dict. Which would limit extensibility in case there are more fields we would like to add.
> 
> I would do that once there are more fields to add, and leave the `dict` but just add a `dataclass`-like container for multiple fields.

I found a reason for using data classes. It makes the typing prettier in the `register` method of the `SktimeRegistry`.

File: sktime/benchmarking/benchmarking_dataclasses.py (hunk 1 above)
Comment (benHeid) [reply sktime/benchmarking/benchmarking_dataclasses.py:129]: As also mentioned somewhere below. Having a class enables us to define methods on the object and thus the code is less complicated. (E.g. to_dataframe here).

File: sktime/benchmarking/benchmarking_dataclasses.py (hunk 2 above)
Comment (benHeid) [reply sktime/benchmarking/benchmarking_dataclasses.py]: Changed to DataFrame

File: sktime/benchmarking/benchmarking_dataclasses.py (hunk 3 above)
Comment (benHeid) [reply sktime/benchmarking/benchmarking_dataclasses.py]: The get_params implementation of the base class?

**Review by felipeangelimvieira — COMMENTED** [review]
Sorry for the delay to provide feedback on this PR Ben. I do not have the full context of the benchmark module, I've just started reading it and have some first impressions. Two questions:

* Why overriding the `run` method from the base class? I see that the BaseBenchmark uses a `_lib_mini_kotsu` module (I didn't know this module before), which I believe the intention is to be something general. I do not see the reason to have a baseclass with an implementation of a public method that is ignored by subclasses. 

* Why validation_id instead of task_id?

**Comment by benHeid** [comment]
One aim is to get rid off kotsu. Overriding run since run is probably the entrypoint off other benchmarks to. 

Validation_id to be backward compatible with the current benchmarking module.

**Review by fkiraly — CHANGES_REQUESTED** [review]
Two questions for review:

- how downwards compatible is this? How is that tested?
- for the docstrings, we need to ensure that newlines surround bullet point lists, or they will not be rendered correctly

**Comment by benHeid** [comment]
> * how downwards compatible is this? How is that tested?

I tested it by leaving the old tests unchanged (some changes are necessary, e.g. removing columns that I added). Thus, it is downward compatible under the assumption that the tests are covering the intended API usage.

> * for the docstrings, we need to ensure that newlines surround bullet point lists, or they will not be rendered correctly


> My other comment chimes with what @felipeangelimvieira says - we should try to retain the generality of the "validator" or "evaluator" from `kotsu`, even if jus tas a function. Also see this PR: #7913 (not advocating for merge, but as an example of how this could be retained)

I do not understand this comment. What is exactly your proposal here? Please note also that until now only the forecasting benchmark has used kotsu. The other benchmarks used something different. So we could move the content of my `run` method into a private method that is called via the run method of the parent class. But I thought the revision was about to get rid off the kotsu based implementation.

**Comment by felipeangelimvieira** [comment]
> I do not understand this comment. What is exactly your proposal here? Please note also that until now only the forecasting benchmark has used kotsu. The other benchmarks used something different. So we could move the content of my run method into a private method that is called via the run method of the parent class. But I thought the revision was about to get rid off the kotsu based implementation.

Would it be possible to remove kotsu with this PR? I think we should either keep one or the other. 
I also think that children of BaseBenchmark should implement a private `_run`, and let BaseBenchmark `_run` empty (raising NotImplementedError) or with a default behaviour (which could be as general as possible). What do you think about this, @benHeid ?

**Comment by benHeid** [comment]
> Would it be possible to remove kotsu with this PR? I think we should either keep one or the other. 

Yes, this PR is removing Kotsu already. And it was one of the goal of this PR to remove all stuff from kotsu. 

> I also think that children of BaseBenchmark should implement a private `_run`, and let BaseBenchmark `_run` empty (raising NotImplementedError) or with a default behaviour (which could be as general as possible). What do you think about this, @benHeid ?

Make sense, this is in easy change. I can add this change.

**Comment by benHeid** [comment]
@fkiraly, as discussed in today's meeting, I did several changes:
* added strategy, and some further parameters to add_task
* deprecated multiple public methods.
  * A special case is BaseBenchmark. I added a deprecation warning, if legacy=True (which is the default value). The reason, why I added this deprecation warning is that the estimators and validations class attributes are public and of a type that I replaced by the new Registry. So I think we need a deprecation cycle here.
* I removed the `_add_task` method of the `BaseBenchmark`, since it was a private method which was doing things that are not needed anymore. Since I removed this method, I also removed three tests in `test_benchmark` that are using this method. Note, `base_benchmark._add_estimator` allowed to pass strings that were converted into task factory methods. This was only possible by using `_add_estimator` of the base class directly. However, the interface of `ForecastingBenchmark.add_estimator` does not supported strings here, it even hasn't the task_entrypoint parameter. So removing it is downward compatible.

**Comment by benHeid** [comment]
After short sync:
* We decided to remove the deprecation cycle of the change in the `__init__` of `BaseBenchmark`, since only by non-intended usage users will notice a change.
* I removed all `kotsu` related files

**Review by fkiraly — APPROVED** [review]
Looks great!

Good basis for getting the module to finally mature.
